proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -2105,32 +3034,1807 @@ lemma doMachineOp_invs_bits[wp]:
+  apply (clarsimp simp: getRefillSize_def2)
+  apply (wpsimp wp: simp: readSchedContext_def)
+  done
+
+lemma refillEmpty_wp:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (scRefillCount ko = 0) s\<rbrace> refillEmpty scp \<lbrace>P\<rbrace>"
+  unfolding refillEmpty_def
+  by (wpsimp wp:)
+
+lemma refillFull_wp:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (scRefillCount ko = scRefillMax ko) s\<rbrace> refillFull scp \<lbrace>P\<rbrace>"
+  unfolding refillFull_def
+  by (wpsimp wp:)
+
+lemma no_ofail_readCurTime[simp]:
+  "no_ofail \<top> readCurTime"
+  unfolding readCurTime_def by clarsimp
+
+lemma ovalid_readCurTime[wp]:
+  "o\<lbrace>\<lambda>s. P (ksCurTime s) s\<rbrace> readCurTime \<lbrace>\<lambda>r s. P r s \<and> r = ksCurTime s\<rbrace>"
+  by (simp add: readCurTime_def asks_def obind_def ovalid_def)
+
+lemma ovalid_readRefillReady[rule_format, simp]:
+  "ovalid (\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (rTime (refillHd ko) \<le> ksCurTime s + kernelWCETTicks) s)
+              (readRefillReady scp) P"
+  unfolding readRefillReady_def readSchedContext_def ovalid_def
+  by (fastforce simp: obind_def split: option.split_asm
+                dest: use_ovalid[OF ovalid_readCurTime]
+               dest!: readObject_misc_ko_at')
+
+
+lemma refillReady_wp:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (rTime (refillHd ko) \<le> ksCurTime s + kernelWCETTicks) s\<rbrace> refillReady scp \<lbrace>P\<rbrace>"
+  unfolding refillReady_def
+  by wpsimp (drule use_ovalid[OF ovalid_readRefillReady])
+
+lemma scActive_wp:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (0 < scRefillMax ko) s\<rbrace> scActive scp \<lbrace>P\<rbrace>"
+  unfolding scActive_def
+  by wpsimp
+
+lemma getRefills_wp:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (scRefills ko) s\<rbrace>
+   getRefills scp
+   \<lbrace>P\<rbrace>"
+  unfolding getRefills_def
+  by wpsimp
+
+lemma refillSufficient_wp:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko scp s \<longrightarrow> P (minBudget \<le> refillsCapacity k (scRefills ko) (scRefillHead ko)) s\<rbrace> refillSufficient scp k \<lbrace>P\<rbrace>"
+  unfolding refillSufficient_def
+  apply (wpsimp wp: getRefills_wp)
+  by (clarsimp simp: sufficientRefills_def obj_at'_def)
+
+(* projection rewrites *)
+
+lemma pred_map_rewrite:
+  "pred_map P proj = opt_pred P proj"
+  by (fastforce simp: pred_map_def2)
+
+abbreviation sc_of2 :: "Structures_A.kernel_object \<rightharpoonup> Structures_A.sched_context" where
+  "sc_of2 ko \<equiv> case ko of kernel_object.SchedContext sc n \<Rightarrow> Some sc | _ \<Rightarrow> None"
+
+abbreviation scs_of2 :: "'z state \<Rightarrow> obj_ref \<rightharpoonup> Structures_A.sched_context" where
+  "scs_of2 \<equiv> (\<lambda>s. kheap s |> sc_of2)"
```

Comment:
This looks like translation between the two projection styles. Not sure there is much we can/should do about this, but it's probably worth taking a moment to make a conscious decision. I naturally tend towards the abbreviation style, because it is more lightweight and I used it in RISCV, but converting all MCS AInvs use to that would be quite a bit of work.

Reply:
Should anything be done about this now, or should it be part of deciding on a heap projection approach?

Reply:
Let's make this part of the heap projection approach decision.